### PR TITLE
[bridge 28/n] redo syncer's curosrs

### PR DESCRIPTION
**Pull request by longbowlu**
## Description 

Both SuiSyncer and EthSyncer query bridge events with cursors. Before this PR,
1. unlike the regular `EventID` cursor (`EventID` = `TransactionDigest` + `EventSeq`), SuiSyncer uses `TransactionDigest`, in the hope that it will simplify the downstream handling. This turns out to be an over optimization and does not work because of how Event cursor is implemented.
2. EthSyncer keeps track of events by block range. But we only update the cursor when there are events. In the case of no event, cursor will stay static. This is not great because eth's RPC only handles up to 50k block ranges' query.

In this PR, we have the following changes:
1. change SuiSyncer's cursor to `EventID`.
2. EthSyncer sends empty event list too, in order to consistently move the cursor forward.

## Test Plan 

unit test cases

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
redo bridge syncer's curosrs.

### CODE REVIEW

File: crates/sui-bridge/src/sui_client.rs [hunk 1]
```diff
@@ -93,100 +93,26 @@ where
...
         package: ObjectID,
         module: Identifier,
-        // Before we support query by checkpoint, we use `tx_digest`` as cursor.
-        // Note the cursor is exclusive, so a callsite passing tx A as cursor
-        // meaning it is intersted in events in transactions after A, globally ordering wise.
-        cursor: TransactionDigest,
-    ) -> BridgeResult<Page<SuiEvent, TransactionDigest>> {
+        // cursor is exclusive
+        cursor: EventID,
+    ) -> BridgeResult<Page<SuiEvent, EventID>> {
```
Comment (longbowlu) [inline crates/sui-bridge/src/sui_client.rs:102]: meat 1 of this PR:
instead of juggling to adapt `TransactionDigest` as the cursor, we simply use the existing RPC `query_events` to do the work.

File: crates/sui-bridge/src/orchestrator.rs [hunk 2]
```diff
@@ -140,12 +138,16 @@ where
...
             Vec<EthLog>,
         )>,
     ) {
         info!("Starting eth watcher task");
-        while let Some((contract, logs)) = eth_events_rx.recv().await {
+        while let Some((contract, end_block, logs)) = eth_events_rx.recv().await {
             if logs.is_empty() {
+                store
+                    .update_eth_event_cursor(contract, end_block)
+                    .expect("Store operation should not fail");
```
Comment (longbowlu) [inline crates/sui-bridge/src/orchestrator.rs:150]: meat 2 of this PR - Eth syncer's cursor is updated even when the event list is empty

File: crates/sui-bridge/src/eth_syncer.rs [hunk 3]
```diff
@@ -156,16 +156,23 @@ where
...
+            events_sender
+                .send((contract_address, end_block, events))
+                .await
+                .expect("All Eth event channel receivers are closed");
+            tracing::info!(
+                ?contract_address,
+                start_block,
+                end_block,
+                "Observed {len} new Eth events",
+            );
```
Comment (longbowlu) [inline crates/sui-bridge/src/eth_syncer.rs:175]: meat 2 of this PR - Eth syncer's sends event list to orchestrator even when the event list is empty

File: crates/sui-bridge/src/config.rs [hunk 4]
```diff
@@ -59,7 +60,7 @@ pub struct BridgeNodeConfig {
     pub eth_bridge_contracts_start_block_override: Option<BTreeMap<String, u64>>,
     /// Override the start transaction digest for each bridge module. Key must be in `sui_bridge_modules`.
     /// When set, SuiSyncer will start from this transaction digest instead of the one in storage.
-    pub sui_bridge_modules_start_tx_override: Option<BTreeMap<String, String>>,
+    pub sui_bridge_modules_start_tx_override: Option<BTreeMap<String, (String, u64)>>,
```
Comment (mwtian) [inline crates/sui-bridge/src/config.rs:63]: Add a comment on what the key and values mean? e.g. is the value tuple a "cursor"?

File: crates/sui-bridge/src/config.rs [hunk 5]
```diff
@@ -226,7 +233,8 @@ pub struct BridgeClientConfig {
     pub eth_bridge_contracts: Vec<EthAddress>,
     pub sui_bridge_modules: Vec<Identifier>,
     pub eth_bridge_contracts_start_block_override: BTreeMap<EthAddress, u64>,
-    pub sui_bridge_modules_start_tx_override: BTreeMap<Identifier, TransactionDigest>,
+    /// The EventID needs to be valid, namely it exists and matches the filter. Otherwise, it will miss one event.
```
Comment (mwtian) [inline crates/sui-bridge/src/config.rs:236]: What does `miss one event` mean?

File: crates/sui-bridge/src/eth_syncer.rs [hunk 6]
```diff
@@ -119,7 +119,7 @@ where
         contract_address: EthAddress,
         mut start_block: u64,
         mut last_finalized_block_receiver: watch::Receiver<u64>,
-        events_sender: mysten_metrics::metered_channel::Sender<(EthAddress, Vec<EthLog>)>,
+        events_sender: mysten_metrics::metered_channel::Sender<(EthAddress, u64, Vec<EthLog>)>,
```
Comment (mwtian) [inline crates/sui-bridge/src/eth_syncer.rs:122]: Maybe block height used for start_block and end_block deserve a new type, instead of just using u64.

File: crates/sui-bridge/src/orchestrator.rs (hunk 2 above)
Comment (mwtian) [reply crates/sui-bridge/src/orchestrator.rs:150]: If the `if logs.is_empty() { ... }` block is removed, `update_eth_event_cursor()` will still be called later, right?

File: crates/sui-bridge/src/sui_client.rs [hunk 7]
```diff
@@ -93,100 +93,26 @@ where
...
         &self,
         package: ObjectID,
         module: Identifier,
-        // Before we support query by checkpoint, we use `tx_digest`` as cursor.
-        // Note the cursor is exclusive, so a callsite passing tx A as cursor
-        // meaning it is intersted in events in transactions after A, globally ordering wise.
-        cursor: TransactionDigest,
-    ) -> BridgeResult<Page<SuiEvent, TransactionDigest>> {
+        // cursor is exclusive
+        cursor: EventID,
```
Comment (mwtian) [inline crates/sui-bridge/src/sui_client.rs:101]: `exclusive_start`?

**Review by mwtian — APPROVED** [review]
Looks good overall

File: crates/sui-bridge/src/config.rs (hunk 4 above)
Comment (lxfind) [inline crates/sui-bridge/src/config.rs:63]: Could you remind us why we could not use the actual types in the config? i.e. EventID and etc.

File: crates/sui-bridge/src/eth_syncer.rs [hunk 8]
```diff
@@ -156,16 +156,23 @@ where
...
-                // are complete per block height. Namely, we should never send a partial list
-                // of events for a block. Otherwise, we may end up missing events.
-                events_sender
-                    .send((contract_address, events))
-                    .await
-                    .expect("All Eth event channel receivers are closed");
-                tracing::info!(?contract_address, "Observed {len} new Eth events",);
-            }
+
+            // Note 1: we send an empty list of events to the channel. This is because of how
```
Comment (lxfind) [inline crates/sui-bridge/src/eth_syncer.rs:160]: Maybe rephrase this to something like: we always send events to the channel even when it's empty..

File: crates/sui-bridge/src/orchestrator.rs (hunk 2 above)
Comment (longbowlu) [reply crates/sui-bridge/src/orchestrator.rs:150]: correct! I put the block here in case we add something in the future that accidentally breaks the logic

File: crates/sui-bridge/src/config.rs (hunk 5 above)
Comment (longbowlu) [reply crates/sui-bridge/src/config.rs:236]: It's how EventID as a cursor works. Today it's implemented in a way that if the given cursor does not exist, or does not match the filter, the returned result will skip one legit event ... which I find inconvenient in many cases. But it's a separate problem to fix cc @emmazzz  @gegaowp

File: crates/sui-bridge/src/eth_syncer.rs (hunk 6 above)
Comment (longbowlu) [reply crates/sui-bridge/src/eth_syncer.rs:122]: good point. I will add a TODO and convert all of them

File: crates/sui-bridge/src/eth_syncer.rs (hunk 8 above)
Comment (longbowlu) [reply crates/sui-bridge/src/eth_syncer.rs:161]: fixed in https://github.com/MystenLabs/sui/pull/15856
